src/CachePeers.cc

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "CachePeers.h"
+#include "SquidConfig.h"
+
+CachePeer *
+CachePeers::nextPeerToPing(const size_t pollIndex)
+{
+    Assure(size());
+
+    // Remember the number of polls to keep shifting each poll starting point,
+    // to avoid always polling the same group of peers before other peers and
+    // risk overloading that first group with requests.
+    if (!pollIndex)
+        ++peerPolls_;
+
+    // subtract 1 to set the very first pos to zero
+    const auto pos = (peerPolls_ - 1 + pollIndex) % size();
+
+    return storage[pos].get();
```

Comment:
This code re-implements the existing `first_ping` feature. This implementation avoids the dangers of having a raw pointer/iterator getting out of sync with its globally-modifiable container.

The code looks a bit strange/unusual because we need to support cache peer scans where each new scan starts with the "next" peer (for load balancing purposes) but iterates through all peers. The start of each scan is detected by zero `pollIndex`. There is similar code in Squid (e.g., storeDirSelectSwapDirRoundRobin()). This implementation avoids caller code duplication and makes the caller code much safer compared with the old ad-hoc approach. It also preserves idiosyncrasies of the related official code -- no `first_ping` functionality changes are expected here.

It is possible to completely refactor these cache peer scans, moving this code outside CachePeers, but those improvements will change the functionality of fairly esoteric Squid features. They are also quite difficult to test. I think they should not be attempted without a pressing need, and definitely not in _this_ PR (that focuses on basic CachePeer storage improvements).

Reply:
IMO we should redesign how all these polling features work and interact. I agree on doing that as a separate project.

For now, please pass around the relevant `Pointer` objects instead of raw-pointers. The whole reason they exist is to avoid/solve the dangers/issues of raw-pointers you appear to be worried about.

Reply:
> please pass around the relevant Pointer objects instead of raw-pointers

This method should not return a pointer at all! Fixed in c01fe3a.